assimilation_code/modules/observations/DEFAULT_obs_kind_mod.F90

```diff
@@ -1386,25 +452,183 @@ function get_index_for_quantity(obs_kind_name)
+! Get string value for quantity by (qty index, item name) of name/value pair
+! Error if qty index out of range
+! Returns '' if no item name match  (or 'null'?)
+
+! (FIXME: terminology: metadata vs item vs ??)
 
-! Accessor function to return observation *type* count
+function get_itemvalue_for_quantity(obs_qty_ind, itemname)
+
+integer, intent(in) :: obs_qty_ind
+character(len=*), intent(in) :: itemname
+character(len=valuelen) :: get_itemvalue_for_quantity
+
+integer :: i
+character(len=*), parameter :: routine = 'get_itemvalue_for_quantity'
+
+if (.not. module_initialized) call initialize_module
+
+call validate_obs_qty_index(obs_qty_ind, routine)
+
+do i=1, obs_qty_info(obs_qty_ind)%nitems
+   if (obs_qty_info(obs_qty_ind)%itemname(i) == itemname) then
+      get_itemvalue_for_quantity = obs_qty_info(obs_qty_ind)%itemvalue(i)
+      return
+   endif
+enddo
+
+get_itemvalue_for_quantity = ''
+
+end function get_itemvalue_for_quantity
+
+!----------------------------------------------------------------------------
+! convenience routine, could also be queried using the generic name=value interface
+! Get bounds values for quantity by index
+! Returns .false. if both min and max are missing_r8
+! A potential code shortcut for calling code if no bounds are present
+
+function has_bounds_for_quantity(obs_qty_ind, minbound, maxbound)
+
+integer,  intent(in)  :: obs_qty_ind
+real(r8), intent(out) :: minbound
+real(r8), intent(out) :: maxbound
+logical :: has_bounds_for_quantity
+
+integer :: i
+character(len=valuelen) :: boundstring
+character(len=*), parameter :: routine = 'get_bounds_for_quantity'
+
+if (.not. module_initialized) call initialize_module
+
+call validate_obs_qty_index(obs_qty_ind, routine)
+
+minbound = MISSING_R8
+maxbound = MISSING_R8
+
+do i=1, obs_qty_info(obs_qty_ind)%nitems
+   if (obs_qty_info(obs_qty_ind)%itemname(i) == "minbound") then
+      boundstring = obs_qty_info(obs_qty_ind)%itemvalue(i)
+      minbound = string_to_real(boundstring)
+   endif
+   if (obs_qty_info(obs_qty_ind)%itemname(i) == "maxbound") then
+      boundstring = obs_qty_info(obs_qty_ind)%itemvalue(i)
+      maxbound = string_to_real(boundstring)
+   endif
+enddo
+
+has_bounds_for_quantity = .not. (minbound == MISSING_R8 .and. maxbound == MISSING_R8)
+
+end function has_bounds_for_quantity
+
+!----------------------------------------------------------------------------
+!> Accessor function to return observation type count
+
+function get_num_types_of_obs()
 
 integer :: get_num_types_of_obs
 
 if ( .not. module_initialized ) call initialize_module
 
-get_num_types_of_obs = max_obs_specific
+get_num_types_of_obs = max_defined_types_of_obs
 
 end function get_num_types_of_obs
 
 !----------------------------------------------------------------------------
+!> Accessor function to return observation quantity count
 
 function get_num_quantities()
 
-! Accessor function to return observation *kind* count
-
 integer :: get_num_quantities
```

Comment:
technically this isn't right.  since quantities start at 0, this is the max valid quantity number.  the number of quantities is this + 1 (to account for the 0 entry).  i wouldn't change it - but if we remove 0 as a valid quantity number then subtleties like this go away.